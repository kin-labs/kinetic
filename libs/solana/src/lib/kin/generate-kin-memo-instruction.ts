import { TransactionInstruction } from '@solana/web3.js'
import { createKinMemo } from './create-kin.memo'
import { generateMemoInstruction } from './generate-memo.instruction'
import { TransactionType } from './index'

export const MEMO_V1_TOKEN_ID = 'Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo'

/**
 * Method to format a correct Kin Memo Instruction based on index and Type
 * @param {number} index
 * @param {TransactionType} type
 * @returns {TransactionInstruction}
 */
export function generateKinMemoInstruction({
  index,
  type,
}: {
  index: number
  type: TransactionType
}): TransactionInstruction {
  // Create correctly formatted memo string, including your App Index
  const content = createKinMemo({ index, type })

  // Create Memo Instruction for KRE Ingestion - Must be Memo Program v1, not v2
  return generateMemoInstruction(content)
}
