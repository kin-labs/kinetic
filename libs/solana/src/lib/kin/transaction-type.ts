export enum TransactionType {
  Unknown = -1,
  None = 0,
  Earn = 1,
  Spend = 2,
  P2P = 3,
}
