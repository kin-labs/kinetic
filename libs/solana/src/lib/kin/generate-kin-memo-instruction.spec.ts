import { generateKinMemoInstruction, MEMO_V1_TOKEN_ID } from './generate-kin-memo-instruction'
import { TransactionType } from './transaction-type'

describe('generateMemoInstruction', () => {
  it('should create a Memo Instruction', async () => {
    const memo = generateKinMemoInstruction({
      index: 1,
      reference: 'ref-type|ref-id',
      type: TransactionType.Spend,
    })

    expect(memo.programId.toBase58()).toEqual(MEMO_V1_TOKEN_ID)
    expect(memo).toMatchSnapshot()
  })
})
