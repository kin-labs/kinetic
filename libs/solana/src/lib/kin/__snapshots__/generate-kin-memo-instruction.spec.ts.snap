// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMemoInstruction should create a Memo Instruction 1`] = `
Object {
  "data": Array [
    82,
    81,
    81,
    65,
    116,
    74,
    55,
    55,
    51,
    54,
    112,
    52,
    116,
    90,
    55,
    55,
    74,
    119,
    73,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    65,
    61,
  ],
  "keys": Array [],
  "programId": "Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo",
}
`;
